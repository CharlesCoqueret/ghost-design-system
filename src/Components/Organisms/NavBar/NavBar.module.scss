.container {
  position: fixed;
  top: 0px;
  width: 100%;
  display: flex;
  align-items: center;
  background-color: rgb(var(--theme-provider-smoke));
  padding: 0;
  margin: 0;
  height: 64px;
  user-select: none;
}

.brand {
  margin: 10px 10px 10px 20px;
  height: auto;

  a {
    all: unset;
    display: flex;
    align-items: center;
  }

  .brandLogo {
    max-height: 22px;
    cursor: pointer;
  }
}

.itemsContainer {
  display: flex;
  justify-content: space-between;
  width: 100%;
}

.itemsGroup {
  display: flex;
  margin: auto 0;
}
