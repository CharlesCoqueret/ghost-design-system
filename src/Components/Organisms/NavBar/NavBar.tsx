import React, { ReactElement } from 'react';
import { NavLink } from 'react-router-dom';
import NavItem, { INavItemProps } from './NavItem';
import SearchBar, { ISearchBarProps } from './SearchBar';

export interface IBrandProps {
  /** Logo positionned at the top right corner of the app */
  logoSource: string;
  /** Link on click on the logo (optional, default: '') */
  redirection?: string;
  /** Alternative to the logo in text format (optional, default: 'logo') */
  alt?: string;
}

export interface INavBarProps {
  /** Brand properties */
  brand: IBrandProps;
  /** Navigation Buttons (on the left side) (optional, default: undefined) */
  navButtons?: Array<INavItemProps>;
  /** Navigation Buttons (on the right side) (optional, default: undefined) */
  navIcons?: Array<INavItemProps>;
  /** Search (on the right side before the navIdons) (optional, default: undefined) */
  searchBar?: ISearchBarProps;
}

const NavBar = (props: INavBarProps): ReactElement => {
  const { brand, navButtons, navIcons, searchBar } = props;

  return (
    <nav className='gds-nav-bar-container noselect'>
      <div className='nav-bar-brand' key='brand'>
        <NavLink to={brand.redirection || NavBar.defaultProps.brand.redirection}>
          <img
            src={brand.logoSource}
            alt={brand.alt ?? NavBar.defaultProps.brand.alt}
            className='nav-bar-brand-image'
          />
        </NavLink>
      </div>
      <div className='nav-bar-items-container' key='items'>
        <div className='nav-bar-items-group' key='leftitems'>
          {navButtons?.map((navItem) => (
            <NavItem
              key={`menu-${
                navItem.label || (navItem.icon && Array.isArray(navItem.icon) && navItem.icon[1]) || navItem.tooltip
              }`}
              {...navItem}
            />
          ))}
        </div>
        <div className='nav-bar-items-group' key='rightitems'>
          {searchBar && <SearchBar {...searchBar} />}
          {navIcons?.map((navItem) => (
            <NavItem key={`menu-${navItem.label}-${JSON.stringify(navItem.icon)}`} {...navItem} />
          ))}
        </div>
      </div>
    </nav>
  );
};

NavBar.defaultProps = {
  brand: { redirection: '/', alt: 'logo' },
  navButtons: undefined,
  navIcons: undefined,
  searchBar: undefined,
};

export default NavBar;
