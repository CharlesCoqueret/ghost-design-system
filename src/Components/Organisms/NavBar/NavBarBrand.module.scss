.brand {
  margin: 10px 10px 10px 20px;
  height: auto;
  display: flex;
  align-items: center;
}

.brandLogo {
  max-height: 22px;
  cursor: pointer;
}
