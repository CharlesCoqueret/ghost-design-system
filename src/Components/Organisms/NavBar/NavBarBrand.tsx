import React, { ReactElement } from 'react';
import { NavLink } from 'react-router-dom';

import styles from './NavBarBrand.module.scss';

export interface IBrandProps {
  /** Logo positionned at the top right corner of the app */
  logoSource: string;
  /** Link on click on the logo (optional, default: '') */
  redirection?: string;
  /** Alternative to the logo in text format (optional, default: 'logo') */
  alt?: string;
}

const NavBarBrand = (props: IBrandProps): ReactElement => {
  const { redirection, logoSource, alt } = props;
  return (
    <NavLink to={redirection || NavBarBrand.defaultProps.redirection} className={styles.brand}>
      <img src={logoSource} alt={alt || NavBarBrand.defaultProps.alt} className={styles.brandLogo} />
    </NavLink>
  );
};

NavBarBrand.defaultProps = {
  redirection: '/',
  alt: 'logo',
};

export default NavBarBrand;
