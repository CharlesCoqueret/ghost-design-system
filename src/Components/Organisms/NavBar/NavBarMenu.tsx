import React, { CSSProperties, PropsWithChildren, ReactElement } from 'react';
import classnames from 'classnames';

import styles from './NavBarMenu.module.scss';

export interface INavBarMenuProps {
  /** Class for the container (optional, default: undefined) */
  className?: string;
  /** Custom style (optional, default: undefined) */
  style?: CSSProperties;
}

const NavBarMenu = (props: PropsWithChildren<INavBarMenuProps>): ReactElement => {
  const { children, className, style } = props;
  return (
    <div className={classnames(styles.group, className)} key='leftitems' style={style}>
      {children}
    </div>
  );
};

export default NavBarMenu;
