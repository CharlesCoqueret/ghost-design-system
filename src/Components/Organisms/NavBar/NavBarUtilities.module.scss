.group {
  display: flex;
  margin: auto 0;
}
