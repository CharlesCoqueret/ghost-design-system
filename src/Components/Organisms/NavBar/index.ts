export type { INavItemProps } from './NavItem';
export { default as NavBarMenu } from './NavBarMenu';
export { default as NavBarUtilities } from './NavBarUtilities';
export { default as NavBar } from './NavBar';
export { default as NavItem } from './NavItem';
