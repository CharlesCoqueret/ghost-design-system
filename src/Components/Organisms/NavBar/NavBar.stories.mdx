import { Canvas, Meta, ArgsTable, Story, Source } from '@storybook/addon-docs';

import NavBar from './NavBar';

import navbarStyle from '../../../assets/_navbar.scss?raw';

export const Template = (args) => {
  return <NavBar {...args} />;
};

<Meta
  title='Organism'
  component={NavBar}
  parameters={{
    viewMode: 'docs',
    layout: 'fullscreen',
  }}
/>

# NavBar

NavBar provides access to destinations and is permanently on-screen.

It can contain:

- Five to seven main destinations (navButtons)
- Five max transversale destinations (navIcons)

Play with NavBar live on [CodeSandbox](https://codesandbox.io/s/sidebar-demo-t4s312)

## Brand

<Canvas isExpanded={true}>
  <NavBar
    brand={{
      logoSource: './images/logo.svg',
      redirection: '/',
    }}
  />
</Canvas>

## Navigation buttons

<Canvas isExpanded={false}>
  <NavBar
    brand={{
      logoSource: './images/logo.svg',
    }}
    navButtons={[
      {
        label: 'Label 1',
        subItems: [
          {
            label: 'Label 1',
            onClick: () => {
              console.log('label 1 - label 1');
            },
          },
          {
            label: 'Label 2',
            onClick: () => {
              console.log('label 1 - label 2');
            },
          },
        ],
      },
      {
        label: 'Label 2',
        subItems: [
          {
            label: 'Label 1',
            onClick: () => {
              console.log('label 2 - label 1');
            },
          },
          {
            label: 'Label 2',
            onClick: () => {
              console.log('label 2 - label 2');
            },
          },
        ],
      },
      {
        label: 'Label 3',
        onClick: () => {
          console.log('label 3');
        },
      },
    ]}
  />
</Canvas>

## NavIcons

<Canvas isExpanded={false}>
  <NavBar
    brand={{
      logoSource: './images/logo.svg',
    }}
    navIcons={[
      {
        icon: ['fal', 'ballot-check'],
        counter: 2,
        subItems: [
          {
            label: 'Label 1',
          },
          {
            label: 'Label 2',
          },
        ],
      },
      {
        icon: ['fal', 'bell'],
        counter: '+99',
        customSubItem: (
          <div
            style={{
              maxWidth: '500px',
              height: '300px',
              padding: '0px 16px',
              overflowX: 'hidden',
              overflowY: 'auto',
              backgroundColor: '#FFF',
            }}>
            <p>
              Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam
              rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate
              sed laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto
              atque quo illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim
              veniam nam odio tempora.
            </p>
            <p>
              Quo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe vel
              consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure
              velit. Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa
              illum ea consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!
            </p>
            <p>
              Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam
              rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate
              sed laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto
              atque quo illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim
              veniam nam odio tempora.
            </p>
            <p>
              Quo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe vel
              consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure
              velit. Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa
              illum ea consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!
            </p>
          </div>
        ),
      },
      {
        icon: ['fal', 'cog'],
        subItems: [
          {
            label: 'Label 1',
          },
          {
            label: 'Label 2',
          },
        ],
      },
      {
        icon: ['fal', 'user-circle'],
        subItems: [
          {
            label: 'Label 1',
          },
          {
            label: 'Label 2',
          },
        ],
      },
    ]}
  />
</Canvas>

## searchBar

<Canvas isExpanded={false}>
  <NavBar
    brand={{
      logoSource: './images/logo.svg',
    }}
    searchBar={{
      placeholder: 'Search',
      onSearch: () => {
        return Promise.resolve([
          <div key='result1'>Result 1</div>,
          <div key='result2'>Result 2</div>,
          <div key='result3'>Result 3</div>,
        ]);
      },
    }}
  />
</Canvas>

## Props

<ArgsTable of={NavBar} sort='requiredFirst' />

For layout management, please refer to this [example](https://codesandbox.io/s/size-management-sidebar-xtpkk0)

## Story

<Canvas>
  <Story
    name='NavBar'
    args={{
      brand: {
        logoSource: './images/logo.svg',
        redirection: '/',
      },
      navButtons: [
        {
          label: 'Label 1',
          subItems: [
            {
              label: 'Label 1',
              onClick: () => {
                console.log('label 1 - label 1');
              },
            },
            {
              label: 'Label 2',
              onClick: () => {
                console.log('label 1 - label 2');
              },
            },
          ],
        },
        {
          label: 'Label 2',
          subItems: [
            {
              label: 'Label 1',
              onClick: () => {
                console.log('label 2 - label 1');
              },
            },
            {
              label: 'Label 2',
              onClick: () => {
                console.log('label 2 - label 2');
              },
            },
          ],
        },
        {
          label: 'Label 3',
          onClick: () => {
            console.log('label 3');
          },
        },
      ],
      searchBar: {
        placeholder: 'Search',
        onSearch: () => {
          return Promise.resolve([
            <div key='result1'>Result 1</div>,
            <div key='result2'>Result 2</div>,
            <div key='result3'>Result 3</div>,
          ]);
        },
      },
      navIcons: [
        {
          icon: ['fal', 'ballot-check'],
          counter: 2,
          subItems: [
            {
              label: 'Label 1',
            },
            {
              label: 'Label 2',
            },
          ],
        },
        {
          icon: ['fal', 'bell'],
          counter: '+99',
          customSubItem: (
            <div
              style={{
                maxWidth: '500px',
                height: '300px',
                padding: '0px 16px',
                overflowX: 'hidden',
                overflowY: 'auto',
                backgroundColor: '#FFF',
              }}>
              <p>
                Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam
                rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum
                cupiditate sed laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui
                architecto atque quo illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem
                unde est enim veniam nam odio tempora.
              </p>
              <p>
                Quo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe vel
                consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure
                velit. Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa
                illum ea consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!
              </p>
              <p>
                Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam
                rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum
                cupiditate sed laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui
                architecto atque quo illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem
                unde est enim veniam nam odio tempora.
              </p>
              <p>
                Quo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe vel
                consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure
                velit. Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa
                illum ea consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!
              </p>
            </div>
          ),
        },
        {
          icon: ['fal', 'cog'],
          subItems: [
            {
              label: 'Label 1',
            },
            {
              label: 'Label 2',
            },
          ],
        },
        {
          icon: ['fal', 'user-circle'],
          subItems: [
            {
              label: 'Label 1',
            },
            {
              label: 'Label very very very very very very very very very very long',
            },
          ],
        },
      ],
    }}>
    {Template.bind({})}
  </Story>
</Canvas>

## Style

[source](https://github.com/CharlesCoqueret/ghost-design-system/blob/main/src/assets/_navbar.scss)

<Source language='css' dark format={true} code={navbarStyle} />
