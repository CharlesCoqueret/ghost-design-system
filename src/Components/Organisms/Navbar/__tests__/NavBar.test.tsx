import React from 'react';
import { render } from '@testing-library/react';

import NavBar from '../Navbar';

describe('NavBar Component', () => {
  it('NavBar renders', async () => {
    const onSearchMock = jest.fn();

    const container = render(
      <NavBar
        brand={{ logoSource: 'logoSource', redirection: 'redirection', alt: 'alt' }}
        navButtons={[{ label: 'button' }]}
        navIcons={[{ label: 'button', icon: ['fal', 'icons'] }]}
        searchBar={{ onSearch: onSearchMock, placeholder: 'placeholder' }}
      />,
    );

    expect(container).toMatchSnapshot();
    expect(onSearchMock).not.toBeCalled();
  });
});
