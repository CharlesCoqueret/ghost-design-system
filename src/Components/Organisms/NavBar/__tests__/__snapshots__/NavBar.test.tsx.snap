// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Component NavBar renders 1`] = `
<div>
  <nav
    class="gds-nav-bar-container noselect"
  >
    <div
      class="nav-bar-brand"
    >
      <a
        href="redirection"
      >
        <img
          alt="alt"
          class="nav-bar-brand-image"
          src="logoSource"
        />
      </a>
    </div>
    <div
      class="nav-bar-items-container"
    >
      <div
        class="nav-bar-items-group"
      >
        <div
          class="nav-bar-menu-item"
        >
          <div
            class="nav-bar-menu-label"
          >
            button
          </div>
        </div>
      </div>
      <div
        class="nav-bar-items-group"
      >
        <div
          class="nav-bar-menu-item nav-bar-search-container"
        >
          <input
            autocomplete="false"
            class="nav-bar-search"
            placeholder="placeholder"
            type="text"
          />
        </div>
        <div>
          <div>
            ControlleMenu
            closed
            has no ref
          </div>
        </div>
        <div
          class="nav-bar-menu-item"
        >
          <i
            class="icon"
            icon="fal,icons"
          />
          <div
            class="nav-bar-menu-label"
          >
            button
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`NavBar Component NavBar renders with variations 1`] = `
<div>
  <nav
    class="gds-nav-bar-container noselect"
  >
    <div
      class="nav-bar-brand"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          class="nav-bar-brand-image"
          src="logosource"
        />
      </a>
    </div>
    <div
      class="nav-bar-items-container"
    >
      <div
        class="nav-bar-items-group"
      >
        <div
          class="nav-bar-menu-item"
        >
          <i
            class="icon"
            icon="fal,icons"
          />
          <div
            class="nav-bar-menu-label"
          >
            button
          </div>
        </div>
      </div>
      <div
        class="nav-bar-items-group"
      >
        <div
          class="nav-bar-menu-item"
        >
          <div
            class="nav-bar-menu-label"
          >
            button
          </div>
        </div>
      </div>
    </div>
  </nav>
</div>
`;

exports[`NavBar Component NavBar renders without buttons or icons 1`] = `
<div>
  <nav
    class="gds-nav-bar-container noselect"
  >
    <div
      class="nav-bar-brand"
    >
      <a
        href="redirection"
      >
        <img
          alt="alt"
          class="nav-bar-brand-image"
          src="logoSource"
        />
      </a>
    </div>
    <div
      class="nav-bar-items-container"
    >
      <div
        class="nav-bar-items-group"
      />
      <div
        class="nav-bar-items-group"
      />
    </div>
  </nav>
</div>
`;
